### PR TITLE
Pl 152 tutorial popup window PL 153 Tutorial popup window question marks, buttons

### PR DESCRIPTION
**Pull request by gabrielekarra**
UI is implemented and everything works
Design is accepted

TODO:
add gif to tutorial windows for highlight

### CODE REVIEW

**Comment by dariomesic** [comment]
sexy +1 by me for the design
for the improvements in future i would make if user clicks on question mark he can move around freely with his mouse and the popup would stay and also background around question mark can be in blue because everything else is hahah

**Review by Giamburrasca — COMMENTED** [review]
Well done +1

**Comment by paolocorsa** [comment]
Perfect!
+1 for me

**Review by NinaGnjidic — COMMENTED** [review]
Wow, this really looks amazing, I could not imagine it looking better!!!!

There are some minor problems that I would like to point out tho... In summary and knowledge graph popup there are typos and grammatical errors which need to be fixed, if you need help with this let us know.
Also I think it would be nice that in those two popups there is some further explaining, for example in summary popup I think that it would be useful if there was an explanation of what type of text these "boxes" hold:
image
And in knowledge graph popup I think it's good to further break down this paragraph in at least three paragraphs (moving, zooming, hovering) in a way that they explain how can user zoom (using mouse scroll), how can user move (on click and drag), and so on...
image
Also, in knowledge graph popup, there should be stated that when depth 2 is selected not all nodes are expanded (due to performance reasons/or some other less lame reason) and that ones that were not expanded can be expanded on click. As well as that, there should be some kind of explanation what depth 1 and depth 2 mean...

File: src/components/PdfReaderComponent.js [hunk 1]
```diff
@@ -163,6 +162,7 @@ class PdfReaderComponent {
     await this.#setCanvasSize();
 
     this.#addVisibilityListenersToPages();
+    this.components.loader.classList.add("hidden");
```
Comment (tonesija) [inline src/components/PdfReaderComponent.js:165]: This should not be here. Function setupPages should not know about some loader html element. This is not the responsibility of setupPages method. Cleaner solution would be to create 'onPdfLoaded' event.

File: src/components/TutorialPopupComponent.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+ * Component representing pop up for every reference and display dynamically the content of it
+ *
+ *
+ * @property {Object} components object that holds DOM elements that represent this component, as well as component's context
+ * @property {HTMLElement} components.questionMarkHighlight question mark for highlight feature tutorial
+ * @property {HTMLElement} components.questionMarkGraph question mark for graph feature tutorial
+ * @property {HTMLElement} components.questionMarkSummary question mark for summary feature tutorial
+ *  @property {HTMLElement} components.panelTutorial div that contains the tutorial window
+ */
+class TutorialPopupComponent {
```
Comment (tonesija) [inline src/components/TutorialPopupComponent.js:11]: Why not separate the question marks in it's own class. That would be cleaner. And then for each question mark you have 1 object of that class.

**Comment by NinaGnjidic** [comment]
Here's what I came up with while fixing the errors in the text:

Highlight commonly used, topic-specific phrasings
A paper can involve using many typical phrases/sentences for introduction, background description and experiment result analyses.
With this tool you can identify and highlight these sentences.
To do it:
Active the "Highlight" toggle in navigation bar.
Scroll the paper and see highlighted sentences.

Knowledge Graph
Create a knowledge graph showing the citation relations of all the papers in the "references section" of the current paper.
This can help you to understand the relations between each work better.
Knowledge graph contains nodes representing papers that can be citing another paper or that can be cited by another paper.
These relationships are visualized by the direction of an arrow on a link between two nodes.
Arrow pointing to node means that paper represented by that node is being cited in paper which is represented by node at the other end of the link.
Click on the knowledge graph button in the navigation bar to generate the knowledge graph. Initially displayed graph has a middle node representing the
paper that is currently being read and the nodes around it represent papers being cited by or citing that paper.
In the knowledge graph window you can:
Zoom in/out on the graph: scroll up/down on your mouse while hovering the graph window.
Move around the graph: click and drag inside the graph window.
Get information about specific node: hover over node and paper title will be displayed over node as well as popup with paper title, authors and fields of study will appear.
Change graph depth: inital graph depth is one, you can click on the slider in the graph window to display depth two which expands some (not all) nodes displayed initially to see even more correlated papers.
Get correlated papers of specific node: click on desired node and it will be expanded with nodes representing the papers correlated to paper represented by that node.
Display the node color legend: hover over a question mark at the bottom right of the graph window and the color legend will appear. Legend explains the connection
between node color and field of study of the paper represented by node. Legened has four pages which can be switched by clicking on circled in the legend.


Summarize your paper
A text wall? Summarize the paper you are reading to just a few sentences!
You can also select a piece of text with the mouse, click on the popup icon that then appears and have that selected text summarized.
Click the summary button in the navigation bar and a side page will be displayed with the following accordion items:
TLDR: "Too Long Didn't Read" - summary of the whole paper in just a few sentences.
ABSTRACT: Summary of the paper or entire thesis.
ABSTRACT SUMMARY: Summary of the abstract of the paper.
SELECTED TEXT SUMMARY: Summary of selected piece of text. When a new text is selected, the content of this summary is accordingly changed.

File: src/components/TutorialPopupComponent.js (hunk 2 above)
Comment (NinaGnjidic) [reply src/components/TutorialPopupComponent.js:14]: I honestly think that these question mark elements should not be here but in ToolbarComponent since their html is within the html of that component. I would then publish events inside mouseenter handlers event per each tutorial window and have a method in this component responsible for displaying just that one window, and have such method for each window. This is the example for _highlight tutorial window_ but can be applied for all other tutorial windows
In #registerEvents method of ToolbarComponent i would add
   this.components.questionMarkHighlight.addEventListener(
      "mouseenter",
      (event) => {
        event.preventDefault();
        EventHandlerService.publish(PDFLEvents.onShowHighlightTutorital) //myb think of better name for this event
      }
    );
    this.components.questionMarkHighlight.addEventListener(
      "mouseleave",
      (event) => {
        event.preventDefault();
        EventHandlerService.publish(PDFLEvents.onHideTutorialWindow) 
      }
    );

and then in #registerEvents method TutorialPopupComponent i would have:
EventHandlerService.subscribe(
      PDFLEvents.onShowHighlightTutorital,
      this.#showHighlightTutorial();
    );
EventHandlerService.subscribe(
      PDFLEvents.onHideTutorialWindow,
      this.#hideTutorialWindow();
    );

Then i would take advantage of "tutorial-window" class which highlight-tutorial-window, graph-tutorial-window and summary-tutorial-window all have and create a method in TutorialPopupComponent called something like cleanWindow which would look something like like:
cleanWindow(){
   document.querySelectorAll(".tutorial-window").forEach(e=>e.classList.add("hidden"));
}
This method can then also be called in #hideTutorialWindow method instead of explicitly hiding elements by id.

Then, in showHighlightTutorial i would call that method and remove hidden just from highlight-tutorial-window element, so something like :
showHighlightTutorial(){
   this.cleanWindow();
   this.components.highlightTutorialWindow.classList.remove("hidden");
}

This way, adding a new tutorial window is easy because you do not have to change logic of existing code, but just add to it and you would also lose the ugly switch case.

Comment (NinaGnjidic) [reply src/components/TutorialPopupComponent.js:14]: Also, you're sometimes referring to this component in both documentation and code as a popup, sometimes as a window and sometimes as panel - choose one and your documentation and code will make far more sense.

**Comment by gabrielekarra** [comment]
@NinaGnjidic all coments fixed (not the one about z-index, but maybe this will be fixed in the task about css). I hade also to delete some parts of the knowledge graph because it was to much big and don't stay in the view (i will add them to tutorial page)

File: src/components/ToolbarComponent.js [hunk 3]
```diff
@@ -51,6 +57,49 @@ class ToolbarComponent {
       this.#showSummaryKey
     );
     this.components.fullScreen.addEventListener("click", this.#showFullScreen);
+
+    this.components.questionMarkHighlight.addEventListener(
+      "mouseenter",
+      (event) => {
+        event.preventDefault();
+        EventHandlerService.publish(PDFLEvents.onShowHighlightTutorialWindow);
```
Comment (matteovisotto) [inline src/components/ToolbarComponent.js:65]: Since the event bus can take one or more parameters after the event name, I suggest (but it can be left as it is) to use a unique event "onShowTutorialWindow" and pass a literal identifier of the tutorial type as a second parameter of the publish. That way is cleaner for two reasons: 1) the event is always the same and change only a property of it (like the shortcuts, it is always a keyup event), 2) The subscription of all tutorial event happens in the same class

Note: the hide event is correct to be separate

**Review by matteovisotto — COMMENTED** [review]
Like it, just the consideration about the event bus.
I'll update this comment when I'll test it

File: src/components/ToolbarComponent.js (hunk 3 above)
Comment (NinaGnjidic) [reply src/components/ToolbarComponent.js:69]: I agree with you but not completely because that way you would still need something like a switch case with **hardcoded strings** like there was before which is better to avoid.

File: src/components/AppComponent/views/ReaderView.js [hunk 4]
```diff
@@ -14,8 +15,9 @@ import { ThumbnailComponent } from "../../ThumbnailComponent.js";
 class ReaderView extends AppView {
   static reader = new PdfReaderComponent();
   static navbar = new NavbarComponent();
+  static tutorial = new TutorialWindowComponent();
```
Comment (NinaGnjidic) [inline src/components/AppComponent/views/ReaderView.js:18]: Not documented.

File: src/components/TutorialWindowComponent.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
+import {
+  EventHandlerService,
+  PDFLEvents,
+} from "../services/EventHandlerService";
+/**
+ * Component representing tutorial window component and it's element
```
Comment (NinaGnjidic) [inline src/components/TutorialWindowComponent.js:6]: This looks like unfinished sentence. You should maybe describe in more detail which tutorials can be displayed in this window (highlight, graph, summary)

File: src/components/TutorialWindowComponent.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    panelTutorial: document.querySelector("#tutorial"),
+    highlightTutorialWindow: document.querySelector(
+      "#highlight-tutorial-window"
+    ),
+    graphTutorialWindow: document.querySelector("#graph-tutorial-window"),
+    summaryTutorialWindow: document.querySelector("#summary-tutorial-window"),
+  };
+
+  /**
+   * Creates the handler service for managing mouse events on the questions marks elements
```
Comment (NinaGnjidic) [inline src/components/TutorialWindowComponent.js:25]: Seems inaccurate

File: src/components/TutorialWindowComponent.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+   * @constructor
+   */
+  constructor() {
+    this.#registerEvents();
+  }
+
+  /**
+   * Adds event listeners to tutorial window component and it's elements.
```
Comment (NinaGnjidic) [inline src/components/TutorialWindowComponent.js:33]: By reading code it looks like only "Adds event listeners to tutorial window component"

**Review by NinaGnjidic — COMMENTED** [review]
Some documentation needs to be worked on, other than that great work here, +1 from me

**Comment by matteovisotto** [comment]
+1 also For me now
